File: bit-manipulation/concatenation-of-consecutive-binary-numbers/concatenation-of-consecutive-binary-numbers.java
class Solution {
  public int concatenatedBinary(int n) {
    long sum = 0;
    int count = 0;
    long mod = (long) 1e9 + 7;

    for (int i = 1; i <= n; i++) {
      if ((i & (i - 1)) == 0) {
        count++;
      }
      sum = ((sum << count) + i) % mod;
    }
    return (int) sum;
  }
}

File: bit-manipulation/concatenation-of-consecutive-binary-numbers/concatenation-of-consecutive-binary-numbers.go
package concatenationofconsecutivebinarynumbers

func concatenatedBinary(n int) int {
	mod := int(1e9 + 7)
	sum := 0
	count := 0
	for i := 1; i <= n; i++ {
		if (i-1)&i == 0 {
			count++
		}
		sum = (sum<<count + i) % mod
	}
	return sum
}
